closed/OpenJ9.gmk

```diff
@@ -164,13 +172,43 @@ endef
 # ----------------------
 # param 1 = The jdk/jre directory name
 # param 2 = The jdk/jre directory to add openj9 content
+ifeq (xr32,$(OPENJ9_PLATFORM_CODE))
+define generated_target_rules
+.PHONY : stage_openj9_$1
+$(foreach file,$(OPENJ9_SHARED_LIBRARIES),$(eval $(call openj9_copy_file,$2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/$(file),$(OUTPUT_ROOT)/vm/$(file))))
+$(foreach file,$(OPENJ9_MISC_FILES) $(OPENJ9_PROPERTY_FILES),$(eval $(call openj9_copy_file,$2/lib/$(file),$(OUTPUT_ROOT)/vm/$(file))))
+$(foreach file,$(OPENJ9_NOTICE_FILES),$(eval $(call openj9_copy_file,$2/$(file),$(SRC_ROOT)/$(file))))
+$(foreach file,$(OPENJ9_REDIRECTOR),$(eval $(call openj9_copy_file,$2/$(OPENJ9_LIBS_OUTPUT_DIR)/j9vm/$(LIBRARY_PREFIX)jvm$(SHARED_LIBRARY_SUFFIX),$(OUTPUT_ROOT)/vm/$(file))))
+# Cross compiling for arm doesn't currently build and stage a build jvm, so the mechanism used to
+# install some (but not all) of the libraries in the jdk and jre images doesn't work.
+# Supplement the existing rules with similar ones for the missing libraries
+$(info adding extra libs for arm cross compile)
+$(foreach file,$(OPENJ9_JAVA_BASE_LIBRARIES),$(eval $(call openj9_copy_file,$2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/$(notdir $(file)),$(OUTPUT_ROOT)/vm/$(file))))
+$(foreach file,$(OPENJ9_SHARED_CLASSES_LIBRARIES),$(eval $(call openj9_copy_file,$2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/$(notdir $(file)),$(OUTPUT_ROOT)/vm/$(file))))
+stage_openj9_java_base_libraries_$1 := $(addprefix $2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/,$(notdir $(OPENJ9_JAVA_BASE_LIBRARIES)))
+stage_openj9_shared_classes_libraries_$1 := $(addprefix $2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/,$(notdir $(OPENJ9_SHARED_CLASSES_LIBRARIES)))
+stage_openj9_shared_libraries_$1 := $(addprefix $2/$(OPENJ9_LIBS_OUTPUT_DIR)/$(LIB_SUBDIR)/,$(OPENJ9_SHARED_LIBRARIES))
+stage_openj9_misc_files_$1 := $(addprefix $2/lib/,$(OPENJ9_MISC_FILES))
+stage_openj9_notice_files_$1 := $(addprefix $2/,$(OPENJ9_NOTICE_FILES))
+stage_openj9_property_files_$1 := $(addprefix $2/lib/,$(OPENJ9_PROPERTY_FILES))
+stage_openj9_redirector_$1 := $2/$(OPENJ9_LIBS_OUTPUT_DIR)/j9vm/$(LIBRARY_PREFIX)jvm$(SHARED_LIBRARY_SUFFIX)
+stage_openj9_$1 : \
+	$$(stage_openj9_java_base_libraries_$1) \
+	$$(stage_openj9_shared_classes_libraries_$1) \
+	$$(stage_openj9_shared_libraries_$1) \
+	$$(stage_openj9_misc_files_$1) \
+	$$(stage_openj9_notice_files_$1) \
+	$$(stage_openj9_redirector_$1) \
+	$$(stage_openj9_property_files_$1)
+endef
+else	# Not an ARM cross compile
```

Comment:
The above looks very similar to what follows. I'd much prefer to avoid this repetition if possible.

Reply:
My first attempt tried to put the conditional just around the extra libraries, but it didn't work. This version is clearly not very elegant, but it's simple to understand and it works. I'd be happy to change it, but my makefile skills are not sufficient.

Reply:
@keithc-ca How strongly do you feel about this? I don't know how to fix it, but I'd be happy to make changes if you show my how.

Reply:
I feel strongly enough to request you give it another try.